content/blog/2023-11-28-context-switching-sprs-on-powerpc.md

```diff
@@ -0,0 +1,443 @@
+Title: Context switching SPRs on PowerPC
+Date: 2023-11-28 08:00:00
+Authors: Benjamin Gray
+Category: Development
+Tags: linux
+
+## Introduction
+
+This article is a dive (well, more of a meander) through some of the PowerPC
+specific aspects of context switching, especially on the Special Purpose
+Register (SPR) handling. It was motivated by my recent work on adding kernel
+support for a hardware feature that interfaces with software through an SPR.
+
+
+## What's a context anyway?
+
+The context we are concerned about in this article is the task context. That's
+all the state that makes up a 'task' in the eyes of the kernel. There are the
+obvious components, like the current register values, thread ID, memory
+mappings, etc., but there are also a few SPRs that get tracked on a per task
+basis. By tracking SPR values for each task context, the kernel can emulate
+per-task support for these SPRs, despite the hardware being per-CPU. The kernel
+simply has to save the values when switching out a task, and set up the values
+stored for the task being switched in.
+
+Unless you're a long time kernel developer, chances are you haven't heard of or
+looked too closely at a 'task' in the kernel. The next section gives a rundown
+of tasks and processes, giving you a better frame of reference for the later
+context switching discussion.
+
+
+## Processes, threads, and tasks
+
+To understand the difference between these three concepts, we'll start with how
+the kernel sees everything: tasks. Tasks are the kernel's view of an executable
+unit, represented by a `struct task_struct`. This is an enormous struct (around
+10K bytes) of all the pieces of data people have wanted to associate with a
+particular unit of execution over the decades. The architecture specific state
+of the task is stored in a one-to-one mapped `struct thread_struct`, available
+through the `thread` member in the `task_struct`. The name 'thread' when
+referring to this structure on a task should not be confused with the concept of
+a thread we'll visit shortly.
+
+A kernel task is highly flexible in terms of resource sharing. Many resources,
+such as the memory mappings and file descriptor tables, are held through
+reference counted handles to a backing data structure. This makes it easy to
+mix-and-match sharing of different components between other tasks.
+
+Crossing over to userspace, here we think of execution in terms of processes and
+threads. Typically, a process is viewed as an execution context that is isolated
```

Comment:
this paragraph doesn't explain the relationship between process/threads and tasks at all